### ISSUE TITLE
Make an API

### ISSUE THREAD

**Issue by tjhorner** [PCMRating/pcmrating#51]
Before I make any drastic changes for an API, I'll explain how I usually make (public) APIs with Rails. It's a lot like how reddit does it:

If you were to go to https://reddit.com/.json right now, you would get JSON data about /r/all.

First, I make an `api_filter` method for each model I want to be able to access from the API. Then, in the controller for that model, in the `show` method, I would put in something like this:

```ruby
def show
  # pretend we get a model here and store it in @model
  respond_to do |type|
    type.json {
      render json: @model.api_filter
    }

    type.html {
      # do normal page stuff here
    }
  end
end
```

...and so on. So if I were to go to http://pcmrating.com/model/32.json it would give me data about a model with ID 32 that can be accessed through the API. Does this sound good? I have an idea for a project and I could integrate this. I don't need to send any data over, only `GET` data.

**Comment by sergiotapia** [comment]
It's better to create a separate base controller for the API, and use something like a serializer to "prepare" the data.

For example:

```
module Api
  module V1
    class BaseController < ActionController::Base
      # something here common with all api endpoints.
    end
  end
end
```

And an resource controller:

```
module Api
  module V1
    class CampaignsController < BaseController
      def latest
        @campaign = Campaign.where('DATE(?) BETWEEN starts_on AND ends_on', Date.today).first
        render json: @campaign
      end

      def show
        @campaign = Campaign.find(params[:id])
        render json: @campaign
      end
    end
  end
end
```

This would go in a folder like: `app/controller/v1/base_controller.rb` - hope this helps you on your way. :) It's a good idea to separate an API like this for versioning and to keep things cleaner.

**Comment by tjhorner** [comment]
It would still be useful to have both, just in case...

In your example, wouldn't this line spit everything about the `Campaign` out?

> `render json: @campaign`

**Comment by sergiotapia** [comment]
Yes, it would spit everything out, unless you configure a serializer. So definitely split up your API into a separate module, and use serializer to make the json nice and cozy for your uses. :smile: 

https://github.com/rails-api/active_model_serializers

**Comment by Ruxton** [comment]
Couldn't agree more with @sergiotapia 

Many past bad experiences have taught me to keep the API in it's own namespace in this situation.  Authentication for the API is different to authentication for the web site.  You only want GET now so you're only entertaining the best way to get that now.

Think further down the track, when people want to interact with the API, not just pull from it.

**Comment by tjhorner** [comment]
Actually, should I keep this open until it actually has an API?

**Comment by sergiotapia** [comment]
Yeah :+1: Keep it open to keep discussion on the API in one place.

**Comment by Lockyy** [comment]
I'm very keen on this. I've been using react.js a bit lately and would love to port the site over eventually, an API would allow that easily.
